pkg/dockerregistry/server/blobdescriptorservice.go

```diff
@@ -25,6 +39,156 @@ type blobDescriptorService struct {
 	distribution.BlobDescriptorService
 }
 
+// Stat returns a a blob descriptor if the given blob is either linked in repository or is referenced in
+// corresponding image stream. This method is invoked from inside of upstream's linkedBlobStore. It expects
+// a proper repository object to be set on given context by upper openshift middleware wrappers.
 func (bs *blobDescriptorService) Stat(ctx context.Context, dgst digest.Digest) (distribution.Descriptor, error) {
-	return dockerRegistry.BlobStatter().Stat(ctx, dgst)
+	repo, found := RepositoryFrom(ctx)
+	if !found || repo == nil {
+		err := fmt.Errorf("failed to retrieve repository from context")
+		context.GetLogger(ctx).Error(err)
+		return distribution.Descriptor{}, err
+	}
+
+	// if there is a repo layer link, return its descriptor
+	desc, err := bs.BlobDescriptorService.Stat(ctx, dgst)
+	if err == nil {
+		// and remember the association
+		repo.cachedLayers.RememberDigest(dgst, repo.blobrepositorycachettl, imageapi.DockerImageReference{
+			Namespace: repo.namespace,
+			Name:      repo.name,
+		}.Exact())
+		return desc, nil
+	}
+
+	context.GetLogger(ctx).Debugf("could not stat layer link %q in repository %q: %v", dgst.String(), repo.Named().Name(), err)
+
+	// verify the blob is stored locally
+	desc, err = dockerRegistry.BlobStatter().Stat(ctx, dgst)
+	if err != nil {
+		return desc, err
+	}
+
+	if imageStreamHasBlob(repo, dgst) {
+		return desc, nil
+	}
+
+	return distribution.Descriptor{}, distribution.ErrBlobUnknown
+}
+
+func (bs *blobDescriptorService) Clear(ctx context.Context, dgst digest.Digest) error {
+	repo, found := RepositoryFrom(ctx)
+	if !found || repo == nil {
+		err := fmt.Errorf("failed to retrieve repository from context")
+		context.GetLogger(ctx).Error(err)
+		return err
+	}
+
```

Comment:
> is this operation accessible via the API?

It is: [`DELETE /v2/<name>/blobs/<digest>`](https://github.com/docker/distribution/blob/master/docs/spec/api.md#deleting-a-layer)

> do we have to guard access to this?

We already do in the auth controller https://github.com/miminar/origin/blob/pullthrough-no-copy-on-head-8613/pkg/dockerregistry/server/auth.go#L370. User needs to be able to `delete images`. But that's not precise. It should be `delete imagestreams/layers`. Non of our client tools invokes that except for pruner though.

> what happens if you clear a given digest that is not actually referenced from a repo or in an image stream? does that let me delete a blob I shouldn't have access to?

This is a wrapper only for linkedBlobStore, which allows to delete only links to blobs. So each invocation may only remove reference to the blob from single repository (aka image stream).